### PR TITLE
Cache warnets on the server

### LINKED ISSUE #155: Cache active networks on server
Avoid recreating them on every call, when not necessary.

### PR DESCRIPTION
**Pull request by willcl-ark**
Avoids repeated rebuilding of the warnet when calling commands

Closes #155

### CODE REVIEW

File: src/warnet/server.py [hunk 1]
```diff
@@ -269,7 +278,8 @@ def network_export(self, network: str) -> str:
         Export all data for sim-ln to subdirectory
         """
         try:
-            wn = Warnet.from_network(network)
+            wn = self.get_warnet(network)
+            self.warnets[network] = wn
```
Comment (pinheadmz) [inline src/warnet/server.py]: why would you need to set this `wn` again?

Comment (willcl-ark) [reply src/warnet/server.py]: That is a good question, I think it's a mistake.

File: src/warnet/server.py [hunk 2]
```diff
@@ -424,7 +434,7 @@ def scenarios_list_running(self) -> list[dict]:
         return running
 
     def network_up(self, network: str = "warnet") -> str:
-        def thread_start(wn):
+        def thread_start(server: Server, network):
```
Comment (pinheadmz) [inline src/warnet/server.py:443]: don't think `server` is used in the function here? Also im not sure but there is a `self.backend` -- i think that `self` *IS* the server?

File: src/warnet/server.py [hunk 3]
```diff
@@ -440,7 +450,8 @@ def thread_start(wn):
 
         try:
             wn = Warnet.from_network(network, self.backend)
-            t = threading.Thread(target=lambda: thread_start(wn))
+            self.warnets[network] = wn
```
Comment (pinheadmz) [inline src/warnet/server.py]: here, too ?

File: src/warnet/server.py [hunk 4]
```diff
@@ -456,11 +467,12 @@ def network_from_file(
         Run a warnet with topology loaded from a <graph_file>
         """
 
-        def thread_start(wn, lock: threading.Lock):
-            with lock:
+        def thread_start(server: Server, network):
```
Comment (pinheadmz) [inline src/warnet/server.py:472]: here too, too. Just not sure why server needs to be passed in

File: src/warnet/server.py [hunk 5]
```diff
@@ -480,11 +492,13 @@ def thread_start(wn, lock: threading.Lock):
...
-            wn = Warnet.from_graph_file(graph_file, config_dir, network, self.backend)
-            t = threading.Thread(target=lambda: thread_start(wn, self.image_build_lock))
+            self.warnets[network] = Warnet.from_graph_file(
+                graph_file, config_dir, network, self.backend
+            )
+            t = threading.Thread(target=lambda: thread_start(self, network))
             t.daemon = True
             t.start()
-            return wn._warnet_dict_representation()
+            return self.warnets[network]._warnet_dict_representation()
```
Comment (pinheadmz) [inline src/warnet/server.py:503]: okay... or you could keep the local `wn` variable and add it to `self.warnets` right after `start()`

**Review by pinheadmz — COMMENTED** [review]
concept ACK, havent tested yet but this is super duper important and badly needed THANK YOU.

One additional comment: if `--force` is passed to `network start` we MUST clear the cached warnet with the same name!

update: okay i think this is already happening in `network_from_file`

**Comment by willcl-ark** [comment]
Thanks for the review, I will re-view it now and incorporate any needed changes.

**Comment by willcl-ark** [comment]
@pinheadmz wdyt of the changes now? I think I addressed all comments

File: src/warnet/server.py (hunk 2 above)
Comment (willcl-ark) [reply src/warnet/server.py:443]: Now using server (I find it makes more sense in a thread than using `self` from the outer scope?)

File: src/warnet/server.py (hunk 5 above)
Comment (willcl-ark) [reply src/warnet/server.py:503]: I think it's the same thing? Just referencing the dict here

Comment (pinheadmz) [reply src/warnet/server.py:503]: it is, its fine thought it looked prettier without the extra dict reference

**Review by pinheadmz — APPROVED** [review]
👍 
🚢 IT

**Pull request merged**